apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/service/state/impl/ShufflingServiceImpl.java

```diff
@@ -0,0 +1,162 @@
+import com.apollocurrency.aplwallet.apl.core.service.state.ShufflingService;
+import com.apollocurrency.aplwallet.apl.util.Listeners;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+@Slf4j
+@Singleton
+public class ShufflingServiceImpl implements ShufflingService {
+
+    private ShufflingDataTable shufflingDataTable;
+    private ShufflingParticipantTable participantTable;
+    private BlockchainConfig blockchainConfig;
+    private TimeService timeService;
+    private Blockchain blockchain;
+
+    private static final Listeners<ShufflingParticipant, ShufflingParticipantEvent> listeners = new Listeners<>();
+
+    @Inject
+    public ShufflingServiceImpl(ShufflingDataTable shufflingDataTable,
+                                ShufflingParticipantTable participantTable,
+                                BlockchainConfig blockchainConfig,
+                                TimeService timeService,
+                                Blockchain blockchain) {
+        this.shufflingDataTable = shufflingDataTable;
+        this.participantTable = participantTable;
+        this.blockchainConfig = blockchainConfig;
+        this.timeService = timeService;
+        this.blockchain = blockchain;
+    }
+
+    @Override
+    public byte[][] getData(long shufflingId, long accountId) {
+        return shufflingDataTable.getData(shufflingId, accountId);
+    }
+
+    @Override
+    public void restoreData(long shufflingId, long accountId, byte[][] data, int timestamp, int height) {
+        shufflingDataTable.restoreData(shufflingId, accountId, data, timestamp, height);
+    }
+
+    @Override
+    public void setData(ShufflingParticipant participant, byte[][] data, int timestamp) {
+        if (data != null && timeService.getEpochTime() - timestamp < blockchainConfig.getMaxPrunableLifetime() && getData(participant.getShufflingId(), participant.getAccountId()) == null) {
+            shufflingDataTable.insert(new ShufflingData(participant.getShufflingId(), participant.getAccountId(), data, timestamp, blockchain.getHeight()));
+        }
+    }
+
+    @Override
+    public DbIterator<ShufflingParticipant> getParticipants(long shufflingId) {
+        return participantTable.getParticipants(shufflingId);
+    }
+
+    @Override
+    public ShufflingParticipant getParticipant(long shufflingId, long accountId) {
+        return participantTable.getParticipant(shufflingId, accountId);
+    }
+
+    @Override
+    public ShufflingParticipant getLastParticipant(long shufflingId) {
+        return participantTable.getLastParticipant(shufflingId);
+    }
+
+    @Override
+    public void addParticipant(long shufflingId, long accountId, int index) {
+        ShufflingParticipant participant = new ShufflingParticipant(shufflingId, accountId, index, blockchain.getHeight());
+        participantTable.addParticipant(participant);
+
+        listeners.notify(participant, ShufflingParticipantEvent.PARTICIPANT_REGISTERED);
```

Comment:
I think so yes, but in any case any not important architecture changes are out of scope of this task. It can takes a lot time to fix bugs after such things.